### DISCUSSION TITLE
Repo level locking in Redis

### DISCUSSION THREAD

**Discussion by astellingwerf** [renovatebot/renovate#11698] [Suggest an Idea]
This may be a wild idea... Renovate allows the use of Redis for its cache. Would it make sense to also use it for repository level locking?

I'm working on a roll out of Renovate for hundreds (possibly thousands) of projects from a single pipeline. I notice that the runs take considerable amounts of time (20 minutes for 49 projects). I've already developed a webhook that listens for events on the MRs and Dependency Dashboard, to selectively retrigger Renovate scanning for the updated projects.

Right now, with a single filesystem cache, this cache will likely grow massive, and to ideally benefit from the cache (and to prevent renovate to interfere on itself) I'll need to serialize the scans. If I'd switch to Redis, the filesystem cache funnel is gone. If we then also store (and read) repo level locking info there, I'd suspect I could significantly decrease the rescan "response time".

My sense of Renovate is that it works on a per-repo basis. Once it starts on one, it will finish it before taking on the next. Is that a correct assumption?

I'd be willing to work on such a feature myself if that helps, but I'd love to hear your ideas on it first.

**Comment by rarkins** [comment]
I would prefer not to add concurrency features to this CLI tool. For now the design of `renovate` is that it has full and exclusive access to the files it uses.

**Reply by astellingwerf** [reply]
Well, the tool would run in the same way, it's just that it would pauze until available for processing or skip and retry later to do whatever it does today. No multithreading would be required.

**Reply by rarkins** [reply]
It will still introduce expectations of concurrency-compatibility which we don't want to set
